### DISCUSSION TITLE
Add support for Tenstorrent Grayskull and Wormhole PCIe cards

### DISCUSSION THREAD

**Discussion by cyrozap** [ggml-org/llama.cpp#10858] [Ideas]
I was going to post a feature request, but the issue template said to post here first. Hopefully I'm not the only one interested in this!

#  Background

[Tenstorrent](https://tenstorrent.com/) makes some AI accelerator cards ([Grayskull](https://tenstorrent.com/hardware/grayskull) and [Wormhole](https://tenstorrent.com/hardware/wormhole)) that connect to a host system over PCIe and are purchasable for reasonable amounts of money (IMO the price would be a lot _more_ reasonable if it was reduced by 50%, but I digress). Probably the most interesting feature of these cards is that the Wormhole variant can be linked to other Wormhole cards over multiple 400G Ethernet links to make one big accelerator with lots of GDDR6 memory--not unlike how Nvidia GPUs can be linked together to share memory over NVLink. Tenstorrent sells some prebuilt systems with this configuration ([TT-LoudBox](https://tenstorrent.com/hardware/tt-loudbox) and [TT-QuietBox](https://tenstorrent.com/hardware/tt-quietbox)), each including four Wormhole n300 cards with a combined total of **96 GB** of device DRAM--enough to run an LLM with 80 billion parameters.

# Request

I have a Wormhole n300 card, and I'd like to get it working with ollama, which means it needs to be working with llama.cpp first. I'm hoping to be able to benchmark it and compare the performance to my 4090.

# Existing efforts

@marty1885 appears to be working on a backend for this already [here](https://github.com/marty1885/llama.cpp/tree/metalium-support), though I'm not sure what the current status of it is.

# Other information

Tenstorrent offers a [cloud service](https://tenstorrent.com/hardware/cloud) with instances that have multiple Grayskull and Wormhole cards. I can't afford to drop $15k on a TT-QuietBox just to find out whether or not the performance is good, but depending on how much the cloud access costs it might be possible to do development and performance testing without breaking the bank.

**Comment by marty1885** [comment]
Hi,
Yes I am working on a backend for Tenstorrent's processors. It is my hobby work but I managed to convince Tenstorrent to give me some hardware to speed things up. We agree that the backend should eventually be upstreamed and Tenstorrent is supportive of the effort. I have some (limited) relation and support from their engineers. Really an awesome bunch of people, specially comparing to other alternative hardware vendors.  The backend is a  work in progress and not ready for production use yet. There's some critical operators missing from TTNN that severally hinders the performance. 

There is also a few problems that is going to make upstreaming challenging, but will deal with that in due time.

> though I'm not sure what the current status of it is.

Documentation is available in the repo. I just wrote it so it's a quite up to date view . https://github.com/marty1885/llama.cpp/blob/metalium-support/docs/backend/Metalium.md

There's a list of op support request in TTNN. Which I hope they can get to soon. https://github.com/tenstorrent/tt-metal/issues?q=is%3Aissue+state%3Aopen+label%3AGGML

> I have a Wormhole n300 card, and I'd like to get it working with ollama, which means it needs to be working with llama.cpp first. I'm hoping to be able to benchmark it and compare the performance to my 4090.

Feel free to try my codebase. But again, it's quite slow as of now. Until the operator support is being added in TTNN. Contributions are totally welcomed. Though IMO most that can be done is done. I spend more time in TTNN fixing their problems lately.

With that said, some notes

* Grayskull will have trouble support some up coming operators due to hardware limitations
* Only smallers models (TinyLLaMA, Gemma 2B) is validated
* FP32 support missing until they fix a few bugs

I'll be speaking at FOSDEM 2025 about the effort. Please come if you happen to be at Brussels. Will love to hear what people think!
https://fosdem.org/2025/schedule/event/fosdem-2025-5122-building-a-new-ggml-backend-how-challenges-and-opportunities-with-novel-accelerators/

---

**Edit 2025-Sep-27:** Now I work for Tenstorrent in their DevRel since May. But my llama.cpp fork maintains a personal project.

**Reply by FlorianHeigl** [reply]
What a great hobby @marty1885 
Hope you enjoyed the chaos that is FOSDEM, i'll go watch your talk now (ain't got a TT card yet, but some day)

**Reply by velonica0** [reply]
> Hi, Yes I am working on a backend for Tenstorrent's processors. It is my hobby work but I managed to convince Tenstorrent to give me some hardware to speed things up. We agree that the backend should eventually be upstreamed and Tenstorrent is supportive of the effort. I have some (limited) relation and support from their engineers. Really an awesome bunch of people, specially comparing to other alternative hardware vendors. The backend is a work in progress and not ready for production use yet. There's some critical operators missing from TTNN that severally hinders the performance.
> 
> There is also a few problems that is going to make upstreaming challenging, but will deal with that in due time.
> 
> > though I'm not sure what the current status of it is.
> 
> Documentation is available in the repo. I just wrote it so it's a quite up to date view . https://github.com/marty1885/llama.cpp/blob/metalium-support/docs/backend/Metalium.md
> 
> There's a list of op support request in TTNN. Which I hope they can get to soon. https://github.com/tenstorrent/tt-metal/issues?q=is%3Aissue+state%3Aopen+label%3AGGML
> 
> > I have a Wormhole n300 card, and I'd like to get it working with ollama, which means it needs to be working with llama.cpp first. I'm hoping to be able to benchmark it and compare the performance to my 4090.
> 
> Feel free to try my codebase. But again, it's quite slow as of now. Until the operator support is being added in TTNN. Contributions are totally welcomed. Though IMO most that can be done is done. I spend more time in TTNN fixing their problems lately.
> 
> With that said, some notes
> 
> * Grayskull will have trouble support some up coming operators due to hardware limitations
> * Only smallers models (TinyLLaMA, Gemma 2B) is validated
> * FP32 support missing until they fix a few bugs
> 
> I'll be speaking at FOSDEM 2025 about the effort. Please come if you happen to be at Brussels. Will love to hear what people think! https://fosdem.org/2025/schedule/event/fosdem-2025-5122-building-a-new-ggml-backend-how-challenges-and-opportunities-with-novel-accelerators/

Hi, @marty1885 

Could you please tell me which version of tt-metal you are using?
I am currently using tt-metal version v0.62.0, and I successfully compiled llama.cpp.
However, running llama-cli results in the error: `pre-allocated tensor (cache_k_l0 (view)) in a buffer (Metalium 0) that cannot run the operation (SET_ROWS)`.

**Reply by marty1885** [reply]
@velonica0 As per `Metalium.md` said - KV Cache has to be stored on the CPU (via the -nkvo flag)

See the document (also linked in my initial response)
https://github.com/marty1885/llama.cpp/blob/metalium-support/docs/backend/Metalium.md

**Reply by velonica0** [reply]
> @velonica0 As per `Metalium.md` said - KV Cache has to be stored on the CPU (via the -nkvo flag)
> 
> See the document (also linked in my initial response) https://github.com/marty1885/llama.cpp/blob/metalium-support/docs/backend/Metalium.md

Sorry, I didn't look at the tutorial after the successful compilation because I was so happy I could use llama.cpp on the T3K.

**Comment by brushknight** [comment]
Thank you for the awesome tutorial on how to put everything together! I just got my n150 and was following along this tutorial and unfortunately, it is not detecting my n150 card. Do you have any steps to troubleshoot what went wrong? I am running proxmox VE with ubuntu guest.

This is what llamacpp first lines look like
```
bin/llama-cli -ngl 23 -m tinyllama-1.1b-chat-v1.0.Q8_0.gguf -p "The solution to Riemann hypothesis is" -nkvo
warning: no usable GPU found, --gpu-layers option will be ignored
warning: one possible reason is that llama.cpp was compiled without GPU support
```

tt-smi shows card normally in the Ubuntu VM
<img width="1770" alt="image" src="https://github.com/user-attachments/assets/d76a2e8d-c2cb-459d-b223-4c54e3fc1be4" />

**Reply by marty1885** [reply]
Few questions

1. Have you set `GGML_METALIUM=ON` in llama.cpp's CMake parameter? My guess is the backend it not compiled
2. To double check, you are using my fork with the correct branch, right?
3. Can you upload your full console output you your command?
4. If you have setup a working TTNN Python environment, can you  run the following script and share the output?

```py
import ttnn

device = ttnn.open_device(device_id=0)
```

Feel free to hop on their discord as I'm quite active there. https://discord.gg/tenstorrent
